# Modelos de ML

In [6]:
import sklearn
print(sklearn.__version__)

1.4.1.post1


In [7]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder
from sklearn.metrics import confusion_matrix, classification_report

#Modelos
from sklearn.linear_model import LogisticRegression

## Read data

In [8]:
df = pd.read_csv('datapreprocess.csv', index_col=[0])
df.head()

edad  abandono             viajes            departamento  distancia_casa  \
id                                                                              
1     41         1      Travel_Rarely                   Sales               1   
2     49         0  Travel_Frequently  Research & Development               8   
4     37         1      Travel_Rarely  Research & Development               2   
5     33         0  Travel_Frequently  Research & Development               3   
7     27         0      Travel_Rarely  Research & Development               2   

        educacion        carrera satisfaccion_entorno implicacion  \
id                                                                  
1   Universitaria  Life Sciences                Media        Alta   
2      Secundaria  Life Sciences                 Alta       Media   
4      Secundaria          Other             Muy_Alta       Media   
5   Universitaria  Life Sciences             Muy_Alta        Alta   
7   Universitaria        Medical                 Baja        Alta   

    nivel_laboral  ... evaluacion satisfaccion_companeros nivel_acciones  \
id                 ...                                                     
1               2  ...       Alta                    Baja              0   
2               2  ...   Muy_Alta                Muy_Alta              1   
4               1  ...       Alta                   Media              0   
5               1  ...       Alta                    Alta              0   
7               1  ...       Alta                Muy_Alta              1   

    anos_experiencia  num_formaciones_ult_ano anos_compania  \
id                                                            
1                  8                        0             6   
2                 10                        3            10   
4                  7                        3             0   
5                  8                        3             8   
7                  6                        3             2   

    anos_desde_ult_promocion anos_con_manager_actual salario_ano  \
id                                                                 
1                          0                       5       71916   
2                          1                       7       61560   
4                          0                       0       25080   
5                          3                       0       34908   
7                          2                       2       41616   

    impacto_abandono  
id                    
1          14670.864  
2          12558.240  
4           4037.880  
5           6876.876  
7           8198.352  

[5 rows x 27 columns]

### Data Preprocessing

In [9]:
#Seleccionamos las variables categoricas
cat = df.select_dtypes('O')

ohe = OneHotEncoder(sparse_output=False)

#entrenamiento y transformacion
cat_ohe = ohe.fit_transform(cat)

#Colocar el nombre a cada columa
cat_ohe = pd.DataFrame(cat_ohe, columns = ohe.get_feature_names_out(input_features=cat.columns)).reset_index(drop = True)
cat_ohe.head()

viajes_Non-Travel  viajes_Travel_Frequently  viajes_Travel_Rarely  \
0                0.0                       0.0                   1.0   
1                0.0                       1.0                   0.0   
2                0.0                       0.0                   1.0   
3                0.0                       1.0                   0.0   
4                0.0                       0.0                   1.0   

   departamento_Human Resources  departamento_Research & Development  \
0                           0.0                                  0.0   
1                           0.0                                  1.0   
2                           0.0                                  1.0   
3                           0.0                                  1.0   
4                           0.0                                  1.0   

   departamento_Sales  educacion_Master  educacion_Primaria  \
0                 1.0               0.0                 0.0   
1                 0.0               0.0                 0.0   
2                 0.0               0.0                 0.0   
3                 0.0               0.0                 0.0   
4                 0.0               0.0                 0.0   

   educacion_Secundaria  educacion_Universitaria  ...  estado_civil_Married  \
0                   0.0                      1.0  ...                   0.0   
1                   1.0                      0.0  ...                   1.0   
2                   1.0                      0.0  ...                   0.0   
3                   0.0                      1.0  ...                   1.0   
4                   0.0                      1.0  ...                   1.0   

   estado_civil_Single  horas_extra_No  horas_extra_Yes  evaluacion_Alta  \
0                  1.0             0.0              1.0              1.0   
1                  0.0             1.0              0.0              0.0   
2                  1.0             0.0              1.0              1.0   
3                  0.0             0.0              1.0              1.0   
4                  0.0             1.0              0.0              1.0   

   evaluacion_Muy_Alta  satisfaccion_companeros_Alta  \
0                  0.0                           0.0   
1                  1.0                           0.0   
2                  0.0                           0.0   
3                  0.0                           1.0   
4                  0.0                           0.0   

   satisfaccion_companeros_Baja  satisfaccion_companeros_Media  \
0                           1.0                            0.0   
1                           0.0                            0.0   
2                           0.0                            1.0   
3                           0.0                            0.0   
4                           0.0                            0.0   

   satisfaccion_companeros_Muy_Alta  
0                               0.0  
1                               1.0  
2                               0.0  
3                               0.0  
4                               1.0  

[5 rows x 49 columns]

In [10]:
#Numeric cols
num = df.select_dtypes('number').reset_index(drop = True)
num.head(1)

edad  abandono  distancia_casa  nivel_laboral  salario_mes  \
0    41         1               1              2         5993   

   num_empresas_anteriores  incremento_salario_porc  nivel_acciones  \
0                        8                       11               0   

   anos_experiencia  num_formaciones_ult_ano  anos_compania  \
0                 8                        0              6   

   anos_desde_ult_promocion  anos_con_manager_actual  salario_ano  \
0                         0                        5        71916   

   impacto_abandono  
0         14670.864

In [11]:
df_ml = pd.concat([cat_ohe, num], axis =1)
df_ml.head()

viajes_Non-Travel  viajes_Travel_Frequently  viajes_Travel_Rarely  \
0                0.0                       0.0                   1.0   
1                0.0                       1.0                   0.0   
2                0.0                       0.0                   1.0   
3                0.0                       1.0                   0.0   
4                0.0                       0.0                   1.0   

   departamento_Human Resources  departamento_Research & Development  \
0                           0.0                                  0.0   
1                           0.0                                  1.0   
2                           0.0                                  1.0   
3                           0.0                                  1.0   
4                           0.0                                  1.0   

   departamento_Sales  educacion_Master  educacion_Primaria  \
0                 1.0               0.0                 0.0   
1                 0.0               0.0                 0.0   
2                 0.0               0.0                 0.0   
3                 0.0               0.0                 0.0   
4                 0.0               0.0                 0.0   

   educacion_Secundaria  educacion_Universitaria  ...  \
0                   0.0                      1.0  ...   
1                   1.0                      0.0  ...   
2                   1.0                      0.0  ...   
3                   0.0                      1.0  ...   
4                   0.0                      1.0  ...   

   num_empresas_anteriores  incremento_salario_porc  nivel_acciones  \
0                        8                       11               0   
1                        1                       23               1   
2                        6                       15               0   
3                        1                       11               0   
4                        9                       12               1   

   anos_experiencia  num_formaciones_ult_ano  anos_compania  \
0                 8                        0              6   
1                10                        3             10   
2                 7                        3              0   
3                 8                        3              8   
4                 6                        3              2   

   anos_desde_ult_promocion  anos_con_manager_actual  salario_ano  \
0                         0                        5        71916   
1                         1                        7        61560   
2                         0                        0        25080   
3                         3                        0        34908   
4                         2                        2        41616   

   impacto_abandono  
0         14670.864  
1         12558.240  
2          4037.880  
3          6876.876  
4          8198.352  

[5 rows x 64 columns]

In [12]:
df_ml.to_csv('dataforml.csv')

### Split dataset
1. X and Y
1. Train and testing set

In [13]:
X = df_ml.drop('abandono', axis = 1)
y = df_ml.abandono
y.value_counts()

abandono
0    1233
1     237
Name: count, dtype: int64

In [14]:
#2. Train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

## Modelos  

#### Modelo GradientBoosting

#### Setup  
Importar Librerias para modelos

In [23]:
!pip install catboost

Defaulting to user installation because normal site-packages is not writeable


In [26]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression, SGDClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, AdaBoostClassifier, ExtraTreesClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.neural_network import MLPClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.metrics import accuracy_score, recall_score, precision_score

In [27]:
#Enternamiento y evaluación de los modelos
models = [
    ("Logistic Regression", LogisticRegression()),
    ("Stochastic Gradient Descent", SGDClassifier()),
    ("Decision Tree Classifier", DecisionTreeClassifier()),
    ("Random Forest Classifier", RandomForestClassifier()),
    ("Gradient Boosting Classifier", GradientBoostingClassifier()),
    ("AdaBoost Classifier", AdaBoostClassifier()),
    ("Extra Trees Classifier", ExtraTreesClassifier()),
    ("SVM", SVC()),
    ("KNeighbors Classifier", KNeighborsClassifier()),
    ("Gaussian Naive Bayes", GaussianNB()),
    ("MLP Classifier", MLPClassifier()),
    ("XGBoost Classifier", XGBClassifier()),
    ("LightGBM Classifier", LGBMClassifier()),
    ("CatBoost Classifier", CatBoostClassifier(silent=True)),
    ("Quadratic Discriminant Analysis", QuadraticDiscriminantAnalysis())
]

results = []
for name, model in models:
    pipeline = Pipeline([
        ("scaler", StandardScaler()),
        ("model", model)
    ])
    pipeline.fit(X_train, y_train)
    y_pred = pipeline.predict(X_test)
    score = accuracy_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    results.append((name, score, recall, precision))

# Print results with both accuracy and recall
for name, accuracy, recall, precision in results:
    print(f"{name}: Accuracy - {accuracy:.4f}, Recall - {recall:.4f}, precision - {precision:.4f}")

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\ensemble\_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
C:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


[LightGBM] [Info] Number of positive: 176, number of negative: 853
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.002498 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1135
[LightGBM] [Info] Number of data points in the train set: 1029, number of used features: 62
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.171040 -> initscore=-1.578276
[LightGBM] [Info] Start training from score -1.578276
Logistic Regression: Accuracy - 0.8617, Recall - 0.3443, precision - 0.5000
Stochastic Gradient Descent: Accuracy - 0.8050, Recall - 0.2787, precision - 0.2881
Decision Tree Classifier: Accuracy - 0.7800, Recall - 0.3607, precision - 0.2750
Random Forest Classifier: Accuracy - 0.8662, Recall - 0.0984, precision - 0.6000
Gradient Boosting Classifier: Accuracy - 0.8662, Recall - 0.2623, precision - 0.5333
AdaBoost Classifier: Accuracy - 0.8617, 

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\discriminant_analysis.py:935: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


In [28]:
#Print Result , Create a DataFrame
df_results = pd.DataFrame(results, columns = ['modelo', 'accuracy', 'recall', 'precision'])
df_results.sort_values(by = 'accuracy', ascending = False)

modelo  accuracy    recall  precision
11               XGBoost Classifier  0.879819  0.344262   0.617647
7                               SVM  0.875283  0.147541   0.750000
13              CatBoost Classifier  0.875283  0.196721   0.666667
3          Random Forest Classifier  0.866213  0.098361   0.600000
4      Gradient Boosting Classifier  0.866213  0.262295   0.533333
6            Extra Trees Classifier  0.866213  0.147541   0.562500
0               Logistic Regression  0.861678  0.344262   0.500000
5               AdaBoost Classifier  0.861678  0.262295   0.500000
8             KNeighbors Classifier  0.857143  0.163934   0.454545
12              LightGBM Classifier  0.857143  0.213115   0.464286
10                   MLP Classifier  0.848073  0.344262   0.437500
1       Stochastic Gradient Descent  0.804989  0.278689   0.288136
2          Decision Tree Classifier  0.780045  0.360656   0.275000
9              Gaussian Naive Bayes  0.671202  0.672131   0.246988
14  Quadratic Discriminant Analysis  0.612245  0.426230   0.160494

#### Hemos Evaluado el:

**Accuracy:** representa el porcentaje total de valores correctamente clasificados, tanto positivos como negativos  

**Recall:** mide la proporción de los casos positivos reales que fueron correctamente identificados por el modelo    

**Precisión:** La métrica de precisión es utilizada para poder saber qué porcentaje de valores que se han clasificado como positivos son realmente positivos



**Observaciones**  

Ya que este dataset posee datos desbalancedos aplicaremos técnicas de balancemiento de datos como son:
- Over Sampling
- Peso en las clases

No utilizare el under sampling ya que la clase *minorítaria tiene 237 registros*

## Técnica de Balancemiento de datos
### Peso en las clases
**Ajuste de Parámetros del modelo:** Consiste en ajustar parametros ó metricas del propio algoritmo para intentar equilibrar a la clase minoritaria penalizando a la clase mayoritaria durante el entrenamiento. Ejemplos on ajuste de peso en árboles, también en logisticregression tenemos el parámetro class_weight= “balanced” que utilizaremos en este ejemplo. No todos los algoritmos tienen estas posibilidades. En redes neuronales por ejemplo podríamos ajustar la métrica de Loss para que penalice a las clases mayoritarias.

In [29]:
#Enternamiento y evaluación de los modelos
models = [
    ("Logistic Regression", LogisticRegression(class_weight='balanced')),
    ("Stochastic Gradient Descent", SGDClassifier(class_weight='balanced')),
    ("Decision Tree Classifier", DecisionTreeClassifier(class_weight='balanced')),
    ("Random Forest Classifier", RandomForestClassifier(class_weight='balanced')),
    #("Gradient Boosting Classifier", GradientBoostingClassifier()),
    #("AdaBoost Classifier", AdaBoostClassifier()),
    ("Extra Trees Classifier", ExtraTreesClassifier(class_weight='balanced')),
    #("SVM", SVC(class_weight='balanced')),
    #("KNeighbors Classifier", KNeighborsClassifier()),
    #("Gaussian Naive Bayes", GaussianNB()),
    #("MLP Classifier", MLPClassifier()),
    #("XGBoost Classifier", XGBClassifier()),
    ("LightGBM Classifier", LGBMClassifier(class_weight='balanced'))
    #("CatBoost Classifier", CatBoostClassifier(silent=True, class_weights='balanced'))
    #("Quadratic Discriminant Analysis", QuadraticDiscriminantAnalysis())
]

results = []
for name, model in models:
    pipeline = Pipeline([
        ("scaler", StandardScaler()),
        ("model", model)
    ])
    pipeline.fit(X_train, y_train)
    y_pred = pipeline.predict(X_test)
    score = accuracy_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    results.append((name, score, recall, precision))

# Print results with both accuracy and recall
for name, accuracy, recall, precision in results:
    print(f"{name}: Accuracy - {accuracy:.4f}, Recall - {recall:.4f}, precision - {precision:.4f}")

[LightGBM] [Info] Number of positive: 176, number of negative: 853
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.002689 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1135
[LightGBM] [Info] Number of data points in the train set: 1029, number of used features: 62
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=-0.000000
[LightGBM] [Info] Start training from score -0.000000
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
Logistic Regression: Accuracy - 0.7324, Recall - 0.6230, precision - 0.2857
Stochastic Gradient Descent: Accuracy - 0.6349, Recall - 0.5574, precision - 0.2024
Decision Tree Classifier: Accuracy - 0.7891, Recall - 0.3770, precision - 0.2949
Random Forest Classifier: Accuracy - 0.8639, Recall - 0.0656, precision - 0.5714
Extra Trees Classifier: Accuracy - 0.8776, Recall - 0.1475, precision - 0.8182
LightGBM Classifier: Accuracy - 0.8

In [30]:
#Print Result , Create a DataFrame
df_p = pd.DataFrame(results, columns = ['modelo', 'accuracy', 'recall', 'precision'])
df_p.sort_values(by = 'accuracy', ascending = False)

modelo  accuracy    recall  precision
4       Extra Trees Classifier  0.877551  0.147541   0.818182
3     Random Forest Classifier  0.863946  0.065574   0.571429
5          LightGBM Classifier  0.843537  0.327869   0.416667
2     Decision Tree Classifier  0.789116  0.377049   0.294872
0          Logistic Regression  0.732426  0.622951   0.285714
1  Stochastic Gradient Descent  0.634921  0.557377   0.202381

**Observaciones**:    
- Para este caso los modelos que se entrenaron con el parametro de penalización de las clases mejoraron el recall, pero disminuyeron el accuracy

## Estrategia: Oversampling de la clase minoritaria
1. Random Over-Sampling (ROS):  
  - Ventaja: Simple y rápido de implementar.
  - Desventaja: Puede generar sobreajuste.
2. SMOTE (Synthetic Minority Over-sampling Technique):  
   - Ventaja: Más sofisticado que ROS, con potencial para una mejor generalización.
  - Desventaja: Puede introducir puntos irreales en la distribución de la clase minoritaria.  
3. ADASYN (Adaptive Synthetic Minority Over-sampling Technique):  
  - Ventaja: Reduce el sobreajuste y mejora la representatividad de la clase minoritaria.
  - Desventaja: Más complejo y computacionalmente costoso que ROS y SMOTE.
4. Borderline-SMOTE (BL-SMOTE):
  - Ventaja: Mejora la clasificación en casos difíciles.
  - Desventaja: Similar a ADASYN en complejidad y costo computacional


**Para este caso utilizare el ADSYN y el BL-SMOTE**

In [31]:
from imblearn.over_sampling import ADASYN

In [32]:
# Definir el over-sampler
adasyn = ADASYN(random_state=42)
# Aplicar over-sampling
X_resampled, y_resampled = adasyn.fit_resample(X, y)

In [33]:
#2. Train/test
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.3, random_state=42)

In [34]:
#Enternamiento y evaluación de los modelos
models = [
    ("Logistic Regression", LogisticRegression()),
    ("Stochastic Gradient Descent", SGDClassifier()),
    ("Decision Tree Classifier", DecisionTreeClassifier()),
    ("Random Forest Classifier", RandomForestClassifier()),
    ("Gradient Boosting Classifier", GradientBoostingClassifier()),
    ("AdaBoost Classifier", AdaBoostClassifier()),
    ("Extra Trees Classifier", ExtraTreesClassifier()),
    ("SVM", SVC()),
    ("KNeighbors Classifier", KNeighborsClassifier()),
    ("Gaussian Naive Bayes", GaussianNB()),
    ("MLP Classifier", MLPClassifier()),
    ("XGBoost Classifier", XGBClassifier()),
    ("LightGBM Classifier", LGBMClassifier()),
    ("CatBoost Classifier", CatBoostClassifier(silent=True)),
    ("Quadratic Discriminant Analysis", QuadraticDiscriminantAnalysis())
]

results = []
for name, model in models:
    pipeline = Pipeline([
        ("scaler", StandardScaler()),
        ("model", model)
    ])
    pipeline.fit(X_train, y_train)
    y_pred = pipeline.predict(X_test)
    score = accuracy_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    results.append((name, score, recall, precision))

# Print results with both accuracy and recall
for name, accuracy, recall, precision in results:
    print(f"{name}: Accuracy - {accuracy:.4f}, Recall - {recall:.4f}, precision - {precision:.4f}")

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\ensemble\_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
C:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


[LightGBM] [Info] Number of positive: 854, number of negative: 849
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.004702 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 5158
[LightGBM] [Info] Number of data points in the train set: 1703, number of used features: 63
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.501468 -> initscore=0.005872
[LightGBM] [Info] Start training from score 0.005872
Logistic Regression: Accuracy - 0.7866, Recall - 0.8040, precision - 0.7602
Stochastic Gradient Descent: Accuracy - 0.7756, Recall - 0.8300, precision - 0.7328
Decision Tree Classifier: Accuracy - 0.8646, Recall - 0.8501, precision - 0.8626
Random Forest Classifier: Accuracy - 0.9316, Recall - 0.8703, precision - 0.9837
Gradient Boosting Classifier: Accuracy - 0.9152, Recall - 0.8617, precision - 0.9553
AdaBoost Classifier: Accuracy - 0.9001, Recall - 0.8646, precision - 0.9202
Extra Trees Classifier: Accura

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\discriminant_analysis.py:935: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


In [35]:
#Print Result , Create a DataFrame
df_adasyn = pd.DataFrame(results, columns = ['modelo', 'accuracy', 'recall', 'precision'])
df_adasyn.sort_values(by = 'accuracy', ascending = False)

modelo  accuracy    recall  precision
6            Extra Trees Classifier  0.939808  0.896254   0.974922
3          Random Forest Classifier  0.931601  0.870317   0.983713
13              CatBoost Classifier  0.926129  0.867435   0.974110
4      Gradient Boosting Classifier  0.915185  0.861671   0.955272
10                   MLP Classifier  0.913817  0.953890   0.875661
12              LightGBM Classifier  0.908345  0.858790   0.943038
11               XGBoost Classifier  0.905609  0.864553   0.931677
7                               SVM  0.904241  0.890490   0.906158
5               AdaBoost Classifier  0.900137  0.864553   0.920245
2          Decision Tree Classifier  0.864569  0.850144   0.862573
0               Logistic Regression  0.786594  0.804035   0.760218
8             KNeighbors Classifier  0.781122  0.956772   0.696017
1       Stochastic Gradient Descent  0.775650  0.829971   0.732824
9              Gaussian Naive Bayes  0.704514  0.821326   0.649203
14  Quadratic Discriminant Analysis  0.704514  0.962536   0.621974

**Comentarios**  
- Observamos una notable mejoria de todos los modelos en cuanto al aumento de sus score en accuracy y recall

### BL-smote

In [36]:
from imblearn.over_sampling import BorderlineSMOTE

In [64]:
# Definir el over-sampler
bl_smote = BorderlineSMOTE(random_state=42)

# Aplicar over-sampling
X_resampled, y_resampled = bl_smote.fit_resample(X, y)
#2. Train/test
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.3, random_state=42)

In [38]:
#Enternamiento y evaluación de los modelos
models = [
    ("Logistic Regression", LogisticRegression()),
    ("Stochastic Gradient Descent", SGDClassifier()),
    ("Decision Tree Classifier", DecisionTreeClassifier()),
    ("Random Forest Classifier", RandomForestClassifier()),
    ("Gradient Boosting Classifier", GradientBoostingClassifier()),
    ("AdaBoost Classifier", AdaBoostClassifier()),
    ("Extra Trees Classifier", ExtraTreesClassifier()),
    ("SVM", SVC()),
    ("KNeighbors Classifier", KNeighborsClassifier()),
    ("Gaussian Naive Bayes", GaussianNB()),
    ("MLP Classifier", MLPClassifier()),
    ("XGBoost Classifier", XGBClassifier()),
    ("LightGBM Classifier", LGBMClassifier()),
    ("CatBoost Classifier", CatBoostClassifier(silent=True)),
    ("Quadratic Discriminant Analysis", QuadraticDiscriminantAnalysis())
]

results = []
for name, model in models:
    pipeline = Pipeline([
        ("scaler", StandardScaler()),
        ("model", model)
    ])
    pipeline.fit(X_train, y_train)
    y_pred = pipeline.predict(X_test)
    score = accuracy_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    results.append((name, score, recall, precision))

# Print results with both accuracy and recall
for name, accuracy, recall, precision in results:
    print(f"{name}: Accuracy - {accuracy:.4f}, Recall - {recall:.4f}, precision - {precision:.4f}")

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\ensemble\_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
C:\ProgramData\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


[LightGBM] [Info] Number of positive: 862, number of negative: 864
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.002579 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 5052
[LightGBM] [Info] Number of data points in the train set: 1726, number of used features: 63
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.499421 -> initscore=-0.002317
[LightGBM] [Info] Start training from score -0.002317
Logistic Regression: Accuracy - 0.8135, Recall - 0.8625, precision - 0.7862
Stochastic Gradient Descent: Accuracy - 0.7135, Recall - 0.7385, precision - 0.7044
Decision Tree Classifier: Accuracy - 0.8703, Recall - 0.8598, precision - 0.8788
Random Forest Classifier: Accuracy - 0.9230, Recall - 0.8841, precision - 0.9591
Gradient Boosting Classifier: Accuracy - 0.9243, Recall - 0.8733, precision - 0.9730
AdaBoost Classifier: Accuracy - 0.9108, Recall - 0.8949, precision - 0.9248
Extra Trees Classifier: Accu

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\discriminant_analysis.py:935: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


In [65]:
#Print Result , Create a DataFrame
df_bl = pd.DataFrame(results, columns = ['modelo', 'accuracy', 'recall', 'precision'])
df_bl.sort_values(by = 'accuracy', ascending = False)

modelo  accuracy    recall  precision
6            Extra Trees Classifier  0.935135  0.908356   0.960114
10                   MLP Classifier  0.925676  0.962264   0.896985
4      Gradient Boosting Classifier  0.924324  0.873315   0.972973
13              CatBoost Classifier  0.924324  0.881402   0.964602
3          Random Forest Classifier  0.922973  0.884097   0.959064
7                               SVM  0.917568  0.916442   0.918919
5               AdaBoost Classifier  0.910811  0.894879   0.924791
12              LightGBM Classifier  0.909459  0.867925   0.947059
11               XGBoost Classifier  0.908108  0.870620   0.941691
2          Decision Tree Classifier  0.870270  0.859838   0.878788
14  Quadratic Discriminant Analysis  0.832432  0.932615   0.777528
8             KNeighbors Classifier  0.816216  0.970350   0.742268
0               Logistic Regression  0.813514  0.862534   0.786241
1       Stochastic Gradient Descent  0.713514  0.738544   0.704370
9              Gaussian Naive Bayes  0.704054  0.865229   0.655102

**Comentarios**  
- Definitivamente esta tecnica mejoro mucho el score de los modelos en cuanto a accuracy y recall

Tomare como modelo el Extra Trees Classifier para predecir con la tecnica BL lo entrenaremos con el dataset completo de X_resample

In [72]:
#setting
best_model = ExtraTreesClassifier(random_state=42, max_depth=5, criterion='entropy')

#entenamiento
best_model.fit(X_resampled, y_resampled)


ExtraTreesClassifier(criterion='entropy', max_depth=5, random_state=42)

In [73]:
#predecir probabilidad
predictions = best_model.predict_proba(X_resampled)[:,1]

In [74]:
predictions

array([0.60344116, 0.28633087, 0.71773905, ..., 0.47207443, 0.52865301,
       0.86382053])

In [79]:
#Evaluation
from sklearn.metrics import roc_auc_score
roc_auc_score(y_resampled, predictions)

0.9543146072884826

### Save  Model

In [80]:
import joblib as jl

In [81]:
jl.dump(best_model, 'trained_model.jl')

['trained_model.jl']

In [82]:
#load model
model = jl.load('trained_model.jl')

#predicciones
pred = model.predict_proba(X_resampled)[:,1]

In [83]:
pred

array([0.60344116, 0.28633087, 0.71773905, ..., 0.47207443, 0.52865301,
       0.86382053])